### PR TITLE
Feat(eos_cli_config_gen): implement global ip dhcp snooping

### LINKED ISSUE #3257: Feat(eos_cli_config_gen): add support for dhcp snooping
### Enhancement summary

Add global config commands for dhcp snooping 
https://arista.my.site.com/AristaCommunity/s/article/dhcp-snooping

### Which component of AVD is impacted

eos_cli_config_gen

### Use case example

```
ip dhcp snooping bridging
ip dhcp snooping information circuit-id type 2 
ip dhcp snooping vlan 1-3900
```
available cli options with eos 4.30.0F
```
access(config)#ip dhcp snooping bridging ?
  <cr>

access(config)#ip dhcp snooping information ?
  option  Configure Information Option

access(config)#ip dhcp snooping information option ?
  circuit-id  Circuit ID subopton in Option-82
  <cr>

access(config)#ip dhcp snooping information option circuit-id ?
  type  Type encoded in the Circuit ID sub-option

access(config)#ip dhcp snooping information option circuit-id type ?
  none     Do not include type in the Circuit ID sub-option
  <0-255>  Specify the type in the Circuit ID sub-option

access(config)#ip dhcp snooping vlan ?
  $         list end
  <1-4094>  VLAN ID or range(s) of VLAN IDs
```

### Describe the solution you would like

```
ip_dhcp_snooping:
  bridging: <bool> (true|false)
  information_option: <bool> (true|false)
    circuit_id_type: <integer> (0-255)
  vlan: <string> (like 10,20,500,1000-2000)
```

### Describe alternatives you have considered

use eos_cli

### Additional context

_No response_

### Contributing Guide

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kmueller68**
## Change Summary

Implement the global configuration options for `ip dhcp snooping` feature.

## Related Issue(s)

Fixes #3257

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes

```
ip_dhcp_snooping:
  enabled: <bool>
  bridging: <bool>
  information_option:
    enabled: <bool>
    circuit_id_type: <str>
    circuit_id_format: <str> 
  vlan: <str>
```

## How to test

```
ip_dhcp_snooping:
  enabled: true
  bridging: true
  information_option:
    enabled: true
    circuit_id_type: none
    circuit_id_format: '%h:%p'
  vlan: "10,20,500,1000-2000"
```

will render the following cli:
```
ip dhcp snooping bridging
ip dhcp snooping information option circuit-id type none format %h:%p
ip dhcp snooping vlan 10,20,500,1000-2000
```

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          enabled:
+            type: bool
+            description: |
+              "Enable insertion of option-82 in DHCP request packets"
+          circuit_id_type:
+            type: str
+            convert_types:
+            - int
+            description: |
+              "< none > or <0 - 255>"
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]:
```suggestion
            description: |
              "none" or <0 - 255>
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+          circuit_id_type_format:
+            type: str
+            convert_types:
+            - int
+            description: |
+              "< %h:%p > Hostname and interface name or
+              < %p:%v> Interface name and VLAN ID"
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]:
```suggestion
            description: |
              - "%h:%p" Hostname and interface name or
              - "%p:%v" Interface name and VLAN ID
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          circuit_id_type:
+            type: str
+            convert_types:
+            - int
+            description: |
+              "< none > or <0 - 255>"
+          circuit_id_type_format:
+            type: str
+            convert_types:
+            - int
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]:
```suggestion
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            type: bool
+            description: |
+              "Enable insertion of option-82 in DHCP request packets"
+          circuit_id_type:
+            type: str
+            convert_types:
+            - int
+            description: |
+              "< none > or <0 - 255>"
+          circuit_id_type_format:
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]: If this option only allows the two described strings, we should add `valid_values` to the schema.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            description: |
+              "< %h:%p > Hostname and interface name or
+              < %p:%v> Interface name and VLAN ID"
+      vlan:
+        type: str
+        convert_types:
+        - int
+        description: |
+          "< vlan_id >, < vlan_id >-< vlan_id >"
+          Example: 15,16,17,18
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]:
```suggestion
        description: |
          VLAN range as string.
          "< vlan_id >, < vlan_id >-< vlan_id >"
          Example: 15,16,17,18
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2 [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+IP DHCP Snooping Insertion of Option 82 is enabled
+{%     endif %}
+{%     if ip_dhcp_snooping.information_option.circuit_id_type is arista.avd.defined %}
+- IP DHCP Snooping Circuit-ID Suboption: {{ ip_dhcp_snooping.information_option.circuit_id_type }}
+{%     endif %}
+{%     if ip_dhcp_snooping.information_option.circuit_id_type.format is arista.avd.defined %}
+- IP DHCP Snooping Circuit-ID Format: {{ ip_dhcp_snooping.information_option.circuit_id_type_format }}
+{%     endif %}
+{%     if ip_dhcp_snooping.vlan is defined and ip_dhcp_snooping.vlan is not none %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2]:
```suggestion
{%     if ip_dhcp_snooping.vlan is arista.avd.defined %}
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2 [hunk 7]
```diff
@@ -0,0 +1,38 @@
+{#
+ Copyright (c) 2023 Arista Networks, Inc.
+ Use of this source code is governed by the Apache License 2.0
+ that can be found in the LICENSE file.
+#}
+{# doc - ip dhcp snooping #}
+{% if ip_dhcp_snooping is defined and ip_dhcp_snooping is not none %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2]:
```suggestion
{% if ip_dhcp_snooping is arista.avd.defined %}
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2 [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+## IP DHCP Snooping
+
+{%     if ip_dhcp_snooping.enabled is arista.avd.defined(true) %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2]:
```suggestion
{%     if ip_dhcp_snooping.enabled is arista.avd.defined(true) %}

```

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2]: Maybe let this `if` surround the whole doc section, so we don't even print the header if it is not enabled. You can just edit the `if` in line 7 to check for `enabled` `true` directly.

Comment (kmueller68) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2]: @ClausHolbechArista Sorry for that inconvenience.
I will go through your comments and try my best.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-dhcp-snooping.j2]: No worries at all. Fix what you can, and I can fix the rest if anything.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml (hunk 4 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]: Done.

**Review by ClausHolbechArista — APPROVED** [review]

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml
Comment (mthiel117) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml:1]: I am not seeing this schema shown in the readthedocs output...

https://ansible-avd--3323.org.readthedocs.build/en/3323/roles/eos_cli_config_gen/docs/input-variables.html#ip-igmp-snooping

**Review by mthiel117 — DISMISSED** [review]
LGTM, but I am not seeing the matching schema for the data model added to the readthdocs output.  

https://ansible-avd--3323.org.readthedocs.build/en/3323/roles/eos_cli_config_gen/docs/input-variables.html#ip-igmp-snooping

Tested Cli on a 720XP with no errors.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml:1]: Good catch. I will fix that.

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml:1]: Fixed in https://github.com/aristanetworks/ansible-avd/pull/3323/commits/811d084cf16a6e79f36aa2bb0bfbe030c6239438

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            type: str
+            convert_types:
+              - int
+            description: |-
+              "none" or <0 - 255>
+          circuit_id_format:
+            type: str
+            description: |-
+              Required if `circuit_id_type` is set.
+              - "%h:%p" Hostname and interface name or
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_dhcp_snooping.schema.yml]:
```suggestion
              - "%h:%p" Hostname and interface name
```

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2 [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+!
+ip dhcp snooping
+{%         endif %}
+{%         if ip_dhcp_snooping.information_option.enabled is arista.avd.defined(true) %}
+{%             if ip_dhcp_snooping.information_option.circuit_id_type is arista.avd.defined and ip_dhcp_snooping.information_option.circuit_id_format is arista.avd.defined %}
+ip dhcp snooping information option circuit-id type {{ ip_dhcp_snooping.information_option.circuit_id_type }} format {{ ip_dhcp_snooping.information_option.circuit_id_format }}
+{%             else %}
+ip dhcp snooping information option
+{%             endif %}
+{%         endif %}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2]: Are we sure for this logic block ?

It seems that when I have
```
host3(config)#show run | grep snoop
ip dhcp snooping bridging
ip dhcp snooping information option circuit-id type 24 format %p:%v
```

only, the option is not active:


```
host3(config)#show ip dhcp snooping
DHCP Snooping is enabled
DHCP Snooping is not operational
DHCP Snooping is configured on following VLANs:
  None
DHCP Snooping bridging is operational on following VLANs:
  None
Insertion of Option-82 is disabled
```


But when I configure:
```
host3(config)#show run | grep snoop
ip dhcp snooping bridging
ip dhcp snooping information option
ip dhcp snooping information option circuit-id type 24 format %p:%v
```


then the option 82 becomes active:
```
host3(config)#show ip dhcp snooping
DHCP Snooping is enabled
DHCP Snooping is operational
DHCP Snooping is configured on following VLANs:
  None
DHCP Snooping bridging is operational on following VLANs:
  None
Insertion of Option-82 is enabled
  Circuit-id default sub-option Type: 24
  Circuit-id default format: Interface name:VLAN ID
  Remote-id: 00:1c:73:64:7b:1f (Switch MAC)
```

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/ip-dhcp-snooping.cfg [hunk 11]
```diff
@@ -0,0 +1,19 @@
+!RANCID-CONTENT-TYPE: arista
+!
+ip dhcp snooping bridging
+ip dhcp snooping information option circuit-id type 10 format %h:%p
+ip dhcp snooping vlan 10,20,500,1000-2000
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/ip-dhcp-snooping.cfg]: I think this should render 

```suggestion
ip dhcp snooping bridging
ip dhcp snooping information option
ip dhcp snooping information option circuit-id type 10 format %h:%p
ip dhcp snooping vlan 10,20,500,1000-2000
```

based on CLI testing (cf comment below) and this TOI: https://www.arista.com/en/support/toi/eos-4-25-2f/14713-dhcp-snooping

**Review by gmuloc — CHANGES_REQUESTED** [review]
Requesting change to block the merge until the question for the additional rendering is solved.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2 (hunk 10 above)
Comment (kmueller68) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2:22]: @gmuloc you are absolutely correct. I justed checked it on a switch right now too.
I am sorry, I missed that in my tests.

```
{#
    Copyright (c) 2023 Arista Networks, Inc.
    Use of this source code is governed by the Apache License 2.0
    that can be found in the LICENSE file.
#}
{# eos - ip dhcp snooping #}
{% if ip_dhcp_snooping is arista.avd.defined %}
{%     if ip_dhcp_snooping.enabled is arista.avd.defined(true) %}
{%         if ip_dhcp_snooping.bridging is arista.avd.defined(true) %}
!
ip dhcp snooping bridging
{%         else %}
!
ip dhcp snooping
{%         endif %}
{%         if ip_dhcp_snooping.information_option.enabled is arista.avd.defined(true) %}
ip dhcp snooping information option
{%             if ip_dhcp_snooping.information_option.circuit_id_type is arista.avd.defined and ip_dhcp_snooping.information_option.circuit_id_format is arista.avd.defined %}
ip dhcp snooping information option circuit-id type {{ ip_dhcp_snooping.information_option.circuit_id_type }} format {{ ip_dhcp_snooping.information_option.circuit_id_format }}
{%             endif %}
{%         endif %}
```

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/ip-dhcp-snooping.cfg (hunk 11 above)
Comment (kmueller68) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/ip-dhcp-snooping.cfg]: @gmuloc you are absolutely correct. I justed checked it on a switch right now too.
I am sorry, I missed that in my tests.

@ClausHolbechArista Would you please be so kind and fix the template for me as bellow.
I am currently struggeling with my environment.

I revised it:
```
{#
    Copyright (c) 2023 Arista Networks, Inc.
    Use of this source code is governed by the Apache License 2.0
    that can be found in the LICENSE file.
#}
{# eos - ip dhcp snooping #}
{% if ip_dhcp_snooping is arista.avd.defined %}
{%     if ip_dhcp_snooping.enabled is arista.avd.defined(true) %}
{%         if ip_dhcp_snooping.bridging is arista.avd.defined(true) %}
!
ip dhcp snooping bridging
{%         else %}
!
ip dhcp snooping
{%         endif %}
{%         if ip_dhcp_snooping.information_option.enabled is arista.avd.defined(true) %}
ip dhcp snooping information option
{%             if ip_dhcp_snooping.information_option.circuit_id_type is arista.avd.defined and ip_dhcp_snooping.information_option.circuit_id_format is arista.avd.defined %}
ip dhcp snooping information option circuit-id type {{ ip_dhcp_snooping.information_option.circuit_id_type }} format {{ ip_dhcp_snooping.information_option.circuit_id_format }}
{%             endif %}
{%         endif %}
{%         if ip_dhcp_snooping.vlan is arista.avd.defined %}
ip dhcp snooping vlan {{ ip_dhcp_snooping.vlan }}
{%         endif %}
{%     endif %}
{% endif %}
```

```
ip_dhcp_snooping:
  enabled: true
  bridging: true
  information_option:
    enabled: true
    circuit_id_type: 10
    circuit_id_format: '%h:%p'
  vlan: "10,20,500,1000-2000
  ```
  
  ```
ip dhcp snooping bridging
ip dhcp snooping information option
ip dhcp snooping information option circuit-id type 10 format %h:%p
ip dhcp snooping vlan 10,20,500,1000-2000
```

test on 720XP:
```
ccess-10-014-01__12:03:09(config)#
access-10-014-01__12:03:09(config)#ip dhcp snooping bridging
access-10-014-01__12:03:53(config)#ip dhcp snooping information option
access-10-014-01__12:03:53(config)#ip dhcp snooping information option circuit-id type 10 format %h:%p
access-10-014-01__12:03:53(config)#ip dhcp snooping vlan 10,20,500,1000-2000
access-10-014-01__12:03:54(config)#
access-10-014-01__12:03:55#show ip dhcp snooping
DHCP Snooping is enabled
DHCP Snooping is operational
DHCP Snooping is configured on following VLANs:
  10,20,500,1000-2000
DHCP Snooping bridging is operational on following VLANs:
  10,20,500,1000-2000
Insertion of Option-82 is enabled
  Circuit-id default sub-option Type: 10
  Circuit-id default format: Hostname:Interface name
  Remote-id: ac:3d:94:c8:0b:10 (Switch MAC)
access-10-014-01__12:04:01#
```

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/ip-dhcp-snooping.cfg]: fixed

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2 (hunk 10 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2]: I have addressed this in an updated template. Please take a look again.

Comment (kmueller68) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2]: @ClausHolbechArista 
I have tested it with your changes again on a 722XPM EOS 4.30.0F successfully:

```
ip_dhcp_snooping:
  enabled: true
  bridging: true
  information_option:
    enabled: true
    circuit_id_type: 10
    circuit_id_format: '%h:%p'
  vlan: "10,20,500,1000-2000"
```

```
access-1h1-01__11:23:14(config)#ip dhcp snooping bridging
access-1h1-01__11:23:43(config)#ip dhcp snooping information option
access-1h1-01__11:23:43(config)#ip dhcp snooping information option circuit-id type 10 format %h:%p
access-1h1-01__11:23:43(config)#ip dhcp snooping vlan 10,20,500,1000-2000
access-1h1-01__11:23:44(config)#show ip dhcp snooping
DHCP Snooping is enabled
DHCP Snooping is operational
DHCP Snooping is configured on following VLANs:
  10,20,500,1000-2000
DHCP Snooping bridging is operational on following VLANs:
  10,20,500,1000-2000
Insertion of Option-82 is enabled
  Circuit-id default sub-option Type: 10
  Circuit-id default format: Hostname:Interface name
  Remote-id: ac:3d:94:c7:d6:ba (Switch MAC)
```

Comment (gmuloc) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ip-dhcp-snooping.j2]: Thanks for checking @kmueller68

**Review by gmuloc — APPROVED** [review]
LGTM with latest changes 🚢

**Comment by gmuloc** [comment]
Thanks @kmueller68 for a great addition - merging it now!

**Comment by kmueller68** [comment]
@gmuloc  @ClausHolbechArista Thank you very much for your support on this!

**Pull request merged**